stackslib/src/core/nonce_cache.rs

```diff
@@ -0,0 +1,253 @@
+            }
+        }
+    }
+
+    pub fn try_flush_with_evicted(
+        &mut self,
+        conn: &mut DBConn,
+        evicted: Option<(StacksAddress, u64)>,
+    ) -> Result<(), db_error> {
+        // Flush the cache to the database
+        let sql = "INSERT OR REPLACE INTO nonces (address, nonce) VALUES (?1, ?2)";
+
+        let tx = conn.transaction()?;
+
+        if let Some((addr, nonce)) = evicted {
+            tx.execute(sql, params![addr, nonce])?;
+        }
+
+        self.cache.flush(|addr, nonce| {
+            tx.execute(sql, params![addr, nonce])?;
+            Ok::<(), db_error>(())
+        })?;
+
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    pub fn flush(&mut self, conn: &mut DBConn) {
+        self.flush_with_evicted(conn, None)
+    }
+}
+
+fn db_set_nonce(conn: &DBConn, address: &StacksAddress, nonce: u64) -> Result<(), db_error> {
+    let addr_str = address.to_string();
+    let nonce_i64 = u64_to_sql(nonce)?;
+
+    let sql = "INSERT OR REPLACE INTO nonces (address, nonce) VALUES (?1, ?2)";
+    conn.execute(sql, params![addr_str, nonce_i64])?;
+    Ok(())
+}
+
+fn db_get_nonce(conn: &DBConn, address: &StacksAddress) -> Result<Option<u64>, db_error> {
+    let addr_str = address.to_string();
+
+    let sql = "SELECT nonce FROM nonces WHERE address = ?";
+    query_row(conn, sql, params![addr_str])
+}
+
+#[cfg(test)]
+mod tests {
+    use clarity::consts::CHAIN_ID_TESTNET;
+    use clarity::types::chainstate::StacksBlockId;
+    use clarity::types::Address;
+    use clarity::vm::tests::{TEST_BURN_STATE_DB, TEST_HEADER_DB};
+
+    use super::*;
+    use crate::chainstate::stacks::db::test::{chainstate_path, instantiate_chainstate};
+    use crate::chainstate::stacks::index::ClarityMarfTrieId;
+    use crate::clarity_vm::clarity::ClarityInstance;
+    use crate::clarity_vm::database::marf::MarfedKV;
+
+    #[test]
+    fn test_nonce_cache() {
+        let _chainstate = instantiate_chainstate(false, 0x80000000, function_name!());
+        let chainstate_path = chainstate_path(function_name!());
+        let mut mempool = MemPoolDB::open_test(false, CHAIN_ID_TESTNET, &chainstate_path).unwrap();
+        let mut cache = NonceCache::new(2);
+
+        let addr1 =
+            StacksAddress::from_string("ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM").unwrap();
+        let addr2 =
+            StacksAddress::from_string("ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5").unwrap();
+        let addr3 =
+            StacksAddress::from_string("ST2CY5V39NHDPWSXMW9QDT3HC3GD6Q6XX4CFRK9AG").unwrap();
+
+        let conn = &mut mempool.db;
+        cache.set(addr1.clone(), 1, conn);
+        cache.set(addr2.clone(), 2, conn);
+
+        let marf = MarfedKV::temporary();
+        let mut clarity_instance = ClarityInstance::new(false, CHAIN_ID_TESTNET, marf);
+        clarity_instance
+            .begin_test_genesis_block(
+                &StacksBlockId::sentinel(),
+                &StacksBlockId([0 as u8; 32]),
```

Comment:
I know you didn't write this and just moved this file, but the proper way to do this is
```suggestion
                &StacksBlockId([0u8; 32]),
```

Reply:
Updated in e4f0b7b8caab984843025aae43a63d3a19721b11.